soundsgood.dsp

```diff
@@ -337,36 +337,83 @@ mscomp8i(target) =
   : ms_dec
 ;
 
-
-Eight_band_Compressor_N_chan(N) =
-   crossover
-  : compressors
-  : mixer
+Eight_band_Compressor_N_chan(N) = si.bus (N) <: si.bus (2 * N)
+  : (calcredux(N,(flist)), si.bus (N))
+  : groupGainOut
+  : par(i, N, shelfcascade ((flist)))
   : outputGain
 with {
-  inputGain = par(i, N, _*inGain);
-  crossover =
-    (
-      (crossoverFreqs<:par(i, Nr_crossoverFreqs, _<:si.bus(N)))
-     ,si.bus(N)
-    )
-    <: ro.interleave(N,Nr_bands)
-    :  par(i, N, fi.crossover8LR4)
-    :  ro.interleave(Nr_bands,N);
-  compressors =
-    (strength_array , thresh_array , att_array , rel_array , knee_array , link_array, ro.interleave(N,Nr_bands))
+
+  /* re-order, interleave output:  left gain [Nr_bands .. 1], left-signal , right gain [Nr_bands .. 1], right-signal
+   * Note: shelfcascade takes gains in reverse order */
+  groupGainOut  =
+    (ro.interleave(Nr_bands,N), si.bus(N))
+    : (ro.interleave(N, Nr_bands+1))
+    : par(i, N, ro.cross(Nr_bands),_);
+
+  /* ******* 8< *******/
+  // TODO: use co.peak_compression_gain_N_chan_db when it arrives in the current faust version
+  peak_compression_gain_mono_db(strength,thresh,att,rel,knee,prePost) =
+    abs : ba.bypass1(prePost,si.onePoleSwitching(att,rel)) : ba.linear2db : gain_computer(strength,thresh,knee) : ba.bypass1((prePost !=1),si.onePoleSwitching(rel,att))
+  with {
+    gain_computer(strength,thresh,knee,level) =
+      select3((level>(thresh-(knee/2)))+(level>(thresh+(knee/2))),
+        0,
+        ((level-thresh+(knee/2)) : pow(2)/(2*max(ma.EPSILON,knee))),
+        (level-thresh))
+      : max(0)*-strength;
+  };
+
+  peak_compression_gain_N_chan_db(strength,thresh,att,rel,knee,prePost,link,1) =
+    peak_compression_gain_mono_db(strength,thresh,att,rel,knee,prePost);
+
+  peak_compression_gain_N_chan_db(strength,thresh,att,rel,knee,prePost,link,N) =
+    par(i, N, peak_compression_gain_mono_db(strength,thresh,att,rel,knee,prePost))
+    <: (si.bus(N),(ba.parallelMin(N) <: si.bus(N))) : ro.interleave(N,2) : par(i,N,(it.interpolate_linear(link)));
+
+  /* ******* >8 *******/
+
+  compressor(N,prePost,strength,thresh,att,rel,knee,link) = peak_compression_gain_N_chan_db (strength,thresh,att,rel,knee,prePost,link,N);
+
+  comps(N) = (strength_array, thresh_array, att_array, rel_array, knee_array, link_array, ro.interleave (N,Nr_bands))
```

Comment:
@x42 Turns out we where both kinda right: the order of the meters was correct in relation to the bands in your version.
However, that is not the order that the compressors need the bands in, in oder to do proper linking of the channels.

To have everything correct, you need my other two suggestions, plus:
```faust
  comps(N) = (strength_array, thresh_array, att_array, rel_array, knee_array, link_array, si.bus(N*Nr_bands))
```

Reply:
Thanks for noticing that. pushed
For some reason I find  `ro.interleave` most confusing.

Reply:
Ah, yeah, I forgot to move the interleave to the end. Thanks.

Perhaps it is easier to use a branch on the main repo and you can continue there (rather than using a fork, since I had no write access)?

Reply:
> Perhaps it is easier to use a branch on the main repo and you can continue there (rather than using a fork, since I had no write access)?

Yeah, how do we do that?
Shall I start a new PR?

Reply:
Cause I just noticed I told you the wrong thing again; there's still a bug.
Working on it...

Reply:
The *original* code (after fixing the `groupGainOut`) worked correctly. I checked by sending sine waves of various freq through it. Now there's indeed something vexxed.

Reply:
> The _original_ code (after fixing the `groupGainOut`) worked correctly. I checked by sending sine waves of various freq through it. Now there's indeed something vexxed.

Except the channel linking was borked. :)

I think I worked it out now, just testing a but more.
Shall I do a new PR when I'm done?

Reply:
BTW, do you find the interpolated "strength, threshold, attack, release and link" parameters useful?
I guess that's also a question for @trummerschlunk - have you done listening tests tweaking those params laid out as-is?

Reply:
> BTW, do you find the interpolated "strength, threshold, attack, release and link" parameters useful? I guess that's also a question for @trummerschlunk - have you done listening tests tweaking those params laid out as-is?

I think it's way better than the alternatives:
- each parameters 8 times
- each parameter once, the same for each compressor
- each parameter once, with hardcoded offsets
- all parameters hardcoded

The first option is great when you want to manually fix certain problems in a mix, but when you are making presets that need to sound good on anything, you want to have the parameters interpolated like this anyway.

We might want to try and have offsets that cater to Fletcher–Munson for the threshold, strength and/or knee.
I am doing [something similar](https://github.com/magnetophon/VoiceOfFaust/blob/82c20ed40fb5f184a236a7c765eab5acfc3bb5f5/lib/general.lib#L249-L277) in Voice of Faust.
With one knob, you can crossfade the bands between no weighting, A-weighting and ITU-R 468 weighting.

@x42 How would you do it?

Reply:
I'd use 
* A common release time for all (no interpolation) bands.
* A single attack value that is however scaled for each freq band (by 1/freq, to be at least 1 period long).
* strengths per band.  -- I think you may not want to to compress bass much (very audible in music),  work on the low and high mids separately for voices (de-ess), and the top also depends. cymbals or ..?
* likewise (relative) thresholds per band.
* knee can be identical for all.
* I don't know how "link" works in a m/s context. It is kind weird, there also comb filter like artifacts when the side limiter kicks in  (generally speaking L/R should be linked, but that doesn't apply here). -- I played with the current controls without being anythewiser while listening.

Reply:
I think the main issue is that there is linear interpolation low -> high.

For dialog, I compress the bass and highs a lot, while leaving the mids alone. For music usually the opposite.
So a "U" shape would work better, and a control (like gamma for brightness) that can turn it into an inverted U?
Also 2 controls only, just not linear interpolation?

Reply:
anyway. let's first focus on getting the mscomp with shelves done and worry about this later. It's not release critical

Reply:
> * A common release time for all (no interpolation) bands.
> * A single attack value that is however scaled for each freq band (by 1/freq, to be at least 1 period long).

You wouldn't want the release time to be relate to the period?
I find I get way cleaner compression with longer release in the lows and shorter in the highs.

The logarithmic interpolation I used here allows for both, with an easy to understand UI.   

> * strengths per band. -- I think you may not want to to compress bass much (very audible in music), work on the low and high mids separately for voices (de-ess), and the top also depends. cymbals or ..?
> * likewise (relative) thresholds per band.

That's a lot of parameters!

>For dialog, I compress the bass and highs a lot, while leaving the mids alone. For music usually the opposite.
So a "U" shape would work better, and a control (like gamma for brightness) that can turn it into an inverted U?
Also 2 controls only, just not linear interpolation?

So for speech an U shape and for music an inverted U?
So for general purpose a line?  ;)

More seriously: For Voice of Faust I also made an interpolator that has a parametric mid: you set the value you want the middle band to have, and you set the number of the band that is the middle.  

> I don't know how "link" works in a m/s context. 

When you unlink you usually mostly compress the middle, cause it's louder, so you get a wider stereo image.
When you link, you keep the stereo image as it is. 

> It is kind weird, there also comb filter like artifacts when the side limiter kicks in 

I haven't heard those. Can you link a track that has this issue with the default settings? 

> I played with the current controls without being anythewiser while listening.

Have you played with the version from this PR or from master?

Reply:
I've mainly used a "made with Ardour" track https://youtu.be/vzdebCx4Kzg
at around 1:50 there are hand-claps + vocals.  Not quite what you get with an applauding audience, but similarly mean. As for m/s this is noticeable on the piano and strings (and reverb).

I did some A/B earlier. The phase issues (fixed by the shelves) is less dramatic than I anticipated.
Even though the song is already mastered it, the loudness increases over time and the range becomes smaller (climax at the end). Also at the end hand-claps are L/R panned and there's a 2nd snare on the right channel.

Reply:
any progress? From what I can tell the earlier version
```
  comps(N) = (strength_array, thresh_array, att_array, rel_array, knee_array, link_array, si.bus(N*Nr_bands))
    : ro.interleave(Nr_bands,6+N)
    : par(i, Nr_bands, compressor(N,prePost)) // : si.bus (N * Nr_bands)
    : par(b, Nr_bands, par(c, N, meter(c+1, b+1)));
 ```
is correct.

Reply:
@x42 Are you still hearing issues with https://github.com/trummerschlunk/soundsgood/pull/38?
I listen to your track on the default settings, and it sounded fine.